lib/internal/modules/esm/get_file_system.js

```diff
@@ -0,0 +1,69 @@
+'use strict';
+
+const {
+  FunctionPrototypeBind,
+  ObjectCreate,
+  ObjectKeys,
+  SafeMap,
+} = primordials;
+
+const realpathCache = new SafeMap();
+
+const internalFS = require('internal/fs/utils');
+const fs = require('fs');
+const fsPromises = require('internal/fs/promises').exports;
+const packageJsonReader = require('internal/modules/package_json_reader');
+const { fileURLToPath } = require('url');
+const { internalModuleStat } = internalBinding('fs');
+
+const implementation = {
+  async readFile(p) {
+    return fsPromises.readFile(p);
+  },
+
+  async statEntry(p) {
+    return internalModuleStat(fileURLToPath(p));
+  },
+
+  async readJson(p) {
+    return packageJsonReader.read(fileURLToPath(p));
+  },
+
+  async realpath(p) {
+    return fsPromises.realpath(p, {
+      [internalFS.realpathCacheKey]: realpathCache
+    });
+  },
+
+  readFileSync(p) {
+    return fs.readFileSync(p);
+  },
+
+  statEntrySync(p) {
+    return internalModuleStat(fileURLToPath(p));
+  },
+
+  readJsonSync(p) {
+    return packageJsonReader.read(fileURLToPath(p));
+  },
+
+  realpathSync(p) {
+    return fs.realpathSync(p, {
+      [internalFS.realpathCacheKey]: realpathCache
+    });
+  },
```

Comment:
In theory realpath can be implemented on top just an `lstat` and `stat` implementation, although that would be a slightly more complex refactoring of the realpath algorithm. It might well simplify the model though in just having three base-level FS primitive hooks to virtualize.

Reply:
I'm a little worried about that - it'd likely be much lower, and the semantic loss could bring unforeseen issues. I'd not feel confident dropping it in this PR 🤔